File: particles/Tutorial/Javascript/Demo/source/TTTRandomComputer.js
/**
 * @license
 * Copyright 2019 Google LLC.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * Code distributed by Google as part of this project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */

/* global defineParticle */

defineParticle(({SimpleParticle}) => class extends SimpleParticle {

  update({gameState, player}) {
    if (!gameState) {
      return;
    }

    if (player && gameState.currentPlayer === player.id) {
      const emptyCells = [];
      const board = JSON.parse(gameState.board);
      // Determine which cells are empty.
      for (let i = 0; i < board.length; i++) {
        if (board[i] === '') {
          emptyCells.push(i);
        }
      }

      const selection = Math.floor(Math.random() * emptyCells.length);
      this.set('computerMove', {move: emptyCells[selection]});
    }
  }
});

File: particles/Tutorial/Kotlin/Demo/src/TTTRandomComputer.kt
/*
 * Copyright 2019 Google LLC.
 *
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 *
 * Code distributed by Google as part of this project is also subject to an additional IP rights
 * grant found at
 * http://polymer.github.io/PATENTS.txt
 */

package arcs.tutorials.tictactoe

import arcs.sdk.wasm.WasmHandle

class TTTRandomComputer : AbstractTTTRandomComputer() {
    override fun onHandleSync(handle: WasmHandle, allSynced: Boolean) = getMove(handles.gameState.fetch())

    init {
        handles.gameState.onUpdate { gameState ->
            getMove(gameState)
        }
    }

    fun getMove(gameState: TTTRandomComputer_GameState?) {
        val gs = gameState ?: TTTRandomComputer_GameState()
        // Ensure we are the current player
        val boardArr = gs.board.split(",")
        val emptyCells = mutableListOf<Double>()

        // Find all the empty cells
        boardArr.forEachIndexed { index, cell ->
            if (cell == "") emptyCells.add(index.toDouble())
        }

        // Choose a random cell as the move
        if (emptyCells.isNotEmpty()) {
            val mv = emptyCells.shuffled().first()
            handles.myMove.set(TTTRandomComputer_MyMove(move = mv))
        }
    }
}
